#15322 Fix/me button plus icon spacing

TrishAllin opened the pull request:
Added right margin to is-compact and is-desktop gridicons within buttons, but exclude that margin on buttons that have gridicon only.

designsimply commented:
I tested and can see the difference here:

![screen shot 2017-07-19 at wed jul 19 1 13 48 pm](https://user-images.githubusercontent.com/1119271/28385193-7080a756-6c84-11e7-9791-6e1433f46355.png)
<sup>Seen at http://calypso.localhost:3000/me using Firefox 54.0.1 on Mac OS X 10.12.5.</sup>

TrishAllin commented:
A pull request #15387 was pushed to fix the spacing on the icon buttons, which has now broken the icon only buttons (see below). Which was why I pushed this pull request that takes in account the icon only button, and I haven't heard anything back about a code check on my solution. @designsimply graciously design reviewed my fix and @kwight has closed the original issue #15322. This is my first pull-request, so I am just making sure that there isn't a protocol I have missed in the documentation that is preventing this from moving forward especially now that a new bug has been introduced? Any help or guidance is appreciated!

![screen shot 2017-07-28 at 9 48 18 am](https://user-images.githubusercontent.com/22596490/28725245-2684fdb2-737a-11e7-8e5d-cc818a2a74c9.png)

kwight commented:
Hi @TrishAllin , thanks for working on this!

If I understand correctly, this PR handled similar issues to #15387 , but better covered them? And now that we've merged #15387 , icon-only buttons are off-center?

Can you update this PR to fix it up, and we'll be happy to get it in :) Thanks!

TrishAllin commented:
That's correct @kwight. You'll notice the trash button is now a bit off. This pull request actually fixes that already (the last/3rd commit recognizes that). I'm excited to be a part of the project now and help out how I can- everyone has been great and it's been a good community experience- thanks!

kwight commented:
Can you match the 4px of #15387 (it had design approval)? Also wondering, is this the only case of icons on their own? I think the screenshot is of the Media controls; are there other instances that would require having the `icon-only` class added? Is there a way to target these situations without adding new classes at all?..

TrishAllin commented:
I will do a run through and try to find all instances of icon only buttons - I looked before but I will be extra vigilant. I used 6px because other icon buttons used that- is 4px more consistent with design standards?

TrishAllin commented:
@kwight I added consideration for some other buttons in the site that were icon-only and the instance where the button has two icons side by side. I also reduced to 4px. (Couldn't really tell what the consistency of icon to text margin was cross site). Let me know what you think. Thanks!

TrishAllin commented:
I found a way to do it without the special class... pushing up a cleaner version. Thanks for pushing for the better solution!

kwight commented on client/components/button/style.scss:
```diff
@@ -63,7 +63,15 @@ button {
+		}
+	}
+
+	&.is-compact  {
```
Extra space in here before the opening bracket. Curious if you've [set up ESLint](https://wpcalypso.wordpress.com/devdocs/docs/coding-guidelines/javascript.md#eslint) in whichever editor you're using? Super helpful for small things like this.

kwight commented:
Sorry, was tough to review this with the GH outage. 🙃

I think this is back to missing the fix for standalone icons? Looks like we still have the right margin:
![screen shot 2017-07-31 at 12 27 14 pm](https://user-images.githubusercontent.com/349751/28794661-d2404dd2-75eb-11e7-99be-81fd230ba4d1.png)

Good work on finding the non-class option! I'm sorry to flip-flop though; if it means needing to add additional unnecessary markup somewhere else, and adding nested selectors, then it's best to stick with your original idea of a simple class. Note that according to the [CSS style guidelines](https://wpcalypso.wordpress.com/devdocs/docs/coding-guidelines/css.md), the class could probably be `.is-icon-only`.

TrishAllin commented:
@kwight No problem! I am getting used to the code base so any help is appreciated. Added back special icon-only class per your advice. I historically have not used a linter for sass... it tends to be a bit aggressive and my sass is generally clean- but I see the value in this project as there are left over style issues from previous contributors. Implemented and fixed.

kwight commented:
Thanks @TrishAllin !

> I historically have not used a linter for sass... it tends to be a bit aggressive and my sass is generally clean

Oh, true, it's not a comment on your quality of code, they just make it easier to ensure consistent code styling across a project with diverse contributors. Like in [this line](https://github.com/Automattic/wp-calypso/pull/16371/files#r130670996), not having spaces around bracket content here is certainly valid (and even more common in many other projects), but isn't consistent with how we've formatted the Calypso project. Setting up a linter allows you to see and fix these things right away, rather than having to wait a round of review.

![screen shot 2017-08-01 at 10 23 44 am](https://user-images.githubusercontent.com/349751/28837964-8d85d492-76a3-11e7-81d0-8d228e04103b.png)

kwight commented:
Hm, does the PR as-is work for you? I'm still seeing the extra `margin-right` like [in the screenshot here](https://github.com/Automattic/wp-calypso/pull/16371#issuecomment-319172984). It looks like it's not kicking in properly, or maybe it's a specificity issue (although I don't see the rule at all, let alone it being overridden in devtools).

Another approach could be passing on the pseudo-selector for a simpler way; having this after the `&.is-compact` declaration around L94 in `client/components/button/style.scss` (to account for specificity) seems to work. What do you think?

```css
&.is-icon-only {
    .gridicon {
        margin-right: 0;
    }
}
```

TrishAllin commented:
No offense was taken at all on the linting ;) On my localhost that extra space is not there and I have a clean workspace where everything I have done has pushed up to this pull request(see below). I just pushed up again to be sure my most current work is there so I'm not sure if something funny is going on with a build or not. Good suggestion to reverse engineer, however, nesting that inside &.is-compact does not account for the &.is-desktop state that is necessary... would we just be trading one evil for the other- having to specify margin separately in the .is-desktop state and .is-compact state and hoping the stack order in .is-compact wouldn't get rearranged and mess up specificity?  I'm imagining it would look like this?
`	&.is-compact {
		padding: 7px;
		color: $gray-text-min;
		font-size: 12px;
		line-height: 1;

		&:disabled {
			color: lighten( $gray, 30% );
		}
		.gridicon {
			top: 4px;
			margin-top: -8px;
                        margin-right: 4px;
		}
                 &.is-icon-only {
                      .gridicon {
                           margin-right: 0;
                      }
                } 
		// Make the left margin of the small plus icon visually less huge
		.gridicons-plus-small {
			margin-left: -4px;
		}
		// Reset the left margin if the button contains only the plus icon
		.gridicons-plus-small:last-of-type {
			margin-left: 0;
		}
		// Make plus icon nudged closer to adjacent icons for add-people and add-plugin type buttons
		.gridicons-plus-small + .gridicon {
			margin-left: -4px;
		}
	}
       &.is-desktop {
           margin-right: 4px;       
       }
`
Let me know your thoughts! Thanks!
![screen shot 2017-08-03 at 11 38 07 am](https://user-images.githubusercontent.com/22596490/28939692-64db3aa2-784f-11e7-841f-a2e54fa94309.png)

TrishAllin commented:
Also, on the note of the linter, when I set up the scss linting (using sublimelinter SublimeLinter-contrib-sass-lint)  it complains about having the spaces in the parens and in digging in the files, some are and some aren't. Should I use spaces in parens or no? (perhaps I don't have the linting set up quite right?) I have altered the button/style.scss file for linter happiness but don't want to push up until I am sure I have the linting set up properly. Thanks!

TrishAllin commented:
I think I have fixed my linter issues. With the extra space in the parens, the only linter errors in button/style.scss I am getting are selector-type-case errors in the commented out headers... does that sound right? Should those be changed to lower case or a different comment format? I wish to appease the linter gods if possible :)
![screen shot 2017-08-03 at 4 51 20 pm](https://user-images.githubusercontent.com/22596490/28947201-e80bbaac-786b-11e7-88ef-e8a800d2a72d.png)

gwwar commented:
@folletto or @rickybanister do you think we could get some design 👀 here on the overall approach and spot check of https://wpcalypso.wordpress.com/devdocs/design/buttons and https://wpcalypso.wordpress.com/devdocs/design/button-group

folletto commented:
This is a very tricky one, because:

1. We pass "any content" to Button components, so there's no programmatic awareness on its content.
2. It's not ideal to specify an extra property that is conditional of content, as it splits the behaviour in two locations (class and content) and makes maintenance less intuitive. It shouldn't be a concern.

Notably, the devdocs simply add a space in front of the text label, which is a fairly straightforward solution, even if I acknowledge it's a bit hacky.
Other places (like `/themes`) simply introduce a specific CSS rule to target a specific button and add padding.

Unfortunately, CSS doesn't allow to recognize "lone text nodes" which means that CSS is not able to separate if it's just an node (gridicon), or a node + text — unless the text is wrapped in a div, which would allow to use adjacency selectors.

This said, my suggestion on the **next steps** would be:

1. **Explore** if there are ways to trigger the CSS rule without any explicit extra in the Button declaration. I can't find in pure CSS (but I might have missed it) but maybe we can do in JS by [checking the passed children](https://www.reactenlightenment.com/basic-react-components/6.8.html)?
2. If we can't find an automated solution like the above, my advice would be to try to avoid splitting the concern, as such I'd rather **wrap the text** in `<label>` or similar when there's an icon. That would be still manual (so inferior to any solution like (1) above) but at least will be just one place to have an extra addition, instead of a class and content. It would be also semantically more accurate (and still just a CSS rule). 
3. Or... we say we are ok just by adding a space? If spaces aren't a viable option, we should **fix the DevDocs** by removing the space because it's the reason we didn't notice until now.

TrishAllin commented:
Thanks @folletto! A few commits back I did explore the css only option with no added class- it did require wrapping the button text in span then applying margin-left to the span. As you probably know there is inconsistency in how the icon + text buttons are handled, some with the text in a span and some without. And you are correct- there is no way to target that text with css. It required me to go to only a few places and wrap the button text. It was easy to do and caused no issues but reverted to adding a class as suggested.
I can definitely pass down a prop "is-icon-only" and apply it that way if that is preferred. I am not sure we can solidly check against the children inside the button though because of the many varied versions- some have one icon, some have two, some have none, some have icon and span some have icon and child.
Either way consistency in button building, be it making sure the icon prop is passed or the architecture of the button itself (icon + span) will need to be maintained. 6 of one half dozen of the other? Let me know if you would like me to submit a prop handled solution instead... I just want to make sure however it is done, it is done the way Calypso prefers/intended :)

folletto commented:
> I can definitely pass down a prop "is-icon-only"

Wait, isn't it what's happening right now? My point is trying to avoid to add anything other than the actual content. :)

So yes, your initial approach by wrapping was imho better than passing something extra (either a class or a prop). I wasn't getting from the discussion what happened — why the wrapping isn't ok? Seems a clean solution to me (unless again we are able to just identify the content as-is).

TrishAllin commented:
Well, it's not really a "prop" being passed down from the button component (like in the case of is-compact), just a hand added class (same difference though for sure). I agree with you completely that wrapping the text and maintaining that button architecture consistency going forward is the cleanest solution (and one that I am used to using in the case of icon + text buttons), and I am not fully sure I understand why reverting to the added class was deemed better in earlier discussion. I am certain there is no way to target the content as-is without too much added complexity. I would be happy to re-implement the icon + span architecture if that is the solution! Should there also be some documentation changed/added somewhere that reflects how icon + text buttons should be created to avoid these issues in the future?

folletto commented:
> I would be happy to re-implement the icon + span architecture if that is the solution! 

That would be my take too. I suggested `label`, but reviewing the spec it seems not a good choice (and was forbidden inside buttons). So `span` should do.

Maybe let's wait until we get a reply from one of the devs, I might be missing the reasoning why a class/prop was better.


> Should there also be some documentation changed/added somewhere that reflects how icon + text buttons should be created to avoid these issues in the future?

Yes. I think fixing the buttons in `devdocs`, and aligning the others you found, will be a strong signal already. We can update the readme too.

gwwar commented:
>  I might be missing the reasoning why a class/prop was better.

In this case I personally find the `is-icon-only` modifier to be a bit odd, since we can easily get into a state where that's set and we set more content 😆 

Anyway, feel free to explore alternative approaches 👍

folletto commented:
Thank you! Let's then go by changing the buttons to use a `<div>` to wrap text too, and have DevDocs match the change. :)

rickybanister commented:
I'm way late to this party, but have perhaps an alternate suggestion.

What if we rely on left margin on the icon and the text to establish the spacing, rather than padding on the button container? That would work if it was icon-only, text-only, or icon-plus-text, wouldn't it? I guess it's a little hacky, but...

kwight commented:
@TrishAllin hi, do the last couple of comments help at all?

apeatling commented:
This is a lot of feedback on what seemed like a trivial change, I think it's good, but I've seen a lot of cases now where it's resulted in something ultimately not getting merged. Something for everyone to think about.

I'm closing this, please reopen if there's new changes, or someone picks it up again.

folletto commented:
> This is a lot of feedback on what seemed like a trivial change, I think it's good, but I've seen a lot of cases now where it's resulted in something ultimately not getting merged. Something for everyone to think about.

Good feedback — but not for this PR.

The button component is one of the most used components in the codebase, and what it seem just a change of a few pixels, it has multiple markup and composition issues as emerged in the discussion above that make the problem far from "trivial".
